async_tcp_server.cpp

```diff
@@ -0,0 +1,116 @@
+#include "async_tcp_server.hpp"
+#include "socket_exception.hpp"
+#include <algorithm>
+#include <cstring>
+#include <iostream>
+#include <sys/select.h>
+
+static const int BUFFER_MAXLEN = 100;
+static const int LISTENQ = 16;
+static const int SERVER_PORT = 8877;
+static const char* ECHO_STR_PREFIX = "echo: ";
+static const int ECHO_STR_LEN = strlen(ECHO_STR_PREFIX);
+
+AsyncTCPServer::AsyncTCPServer() 
+{
+    Socket_sptr master_socket = std::make_shared<Socket>(AF_INET, SOCK_STREAM, 0);
+
+    m_async_engine = std::make_unique<AsyncEngine>(master_socket);
+}
+
+
+AsyncTCPServer::~AsyncTCPServer() 
+{
+
+}
+
+
+void AsyncTCPServer::SendEcho(Socket_sptr active_socket, const char* buffer, const int buffer_size)
+{
+    std::cout << "received " << buffer << "from " <<
+    active_socket->GetIPAddressStr() << ":" << active_socket->GetPort() << std::endl;
+    std::string echo_str("echo: ");
+    echo_str.append(buffer);
+    active_socket->Send(echo_str.c_str(), echo_str.length(), 0);
+    memset((void*)buffer, 0, buffer_size);
+}
+
+void AsyncTCPServer::RunAsyncServer(const int server_port)
+{
+    try 
+    {
+        m_async_engine->GetMasterSocket()->Bind(INADDR_ANY, server_port);
+        m_async_engine->GetMasterSocket()->Listen(LISTENQ);
+
+        while(true)
+        {
+            Socket_vector ready_sockets = m_async_engine->Select(std::chrono::milliseconds(1300));
+            
+            for(auto socket : ready_sockets)
+            {
+                if(socket == m_async_engine->GetMasterSocket())
+                {
+                    AcceptIncomingConnection();
+                }
+                else
+                {
+                    char buffer[BUFFER_MAXLEN];
+                    memset(buffer, 0, BUFFER_MAXLEN);
+
+                    int received_bytes_count = socket->Recv(buffer, sizeof(buffer), 0);
+
+                    if(received_bytes_count == 0)
+                    {
+                        CloseConnection(socket);
+                    }
+                    else
+                    {
+                        SendEcho(socket, buffer, BUFFER_MAXLEN);
+                    }
+                }
+            }
+        }
+
+    }
+    catch (SocketException & se)
+    {
+        std::cout << se.what() << std::endl;
+    }
+    catch (std::runtime_error & re)
+    {
+        std::cout << re.what() << std::endl;
+    }
+}
+
+Socket_sptr AsyncTCPServer::AcceptIncomingConnection()
```

Comment:
Seems like there is no reason for this function to return accepted socket, it is already added to async engine anyway.

Reply:
BTW, it does not fully conform single responsibility principle, but let it be on your conscience. ;)